### PR TITLE
[NeMo-UX] Dataloading enhancements and bug fixes

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

- Adds support for passing multiple paths to `PreTrainingDataModule`
- Provides support for resuming dataloader when restoring from a checkpoint

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by marcromeyn — APPROVED** [review]

**Review by marcromeyn — DISMISSED** [review]

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 1]
```diff
@@ -133,7 +142,7 @@ def gpt_dataset_config(self) -> "GPTDatasetConfig":
         from megatron.core.datasets.gpt_dataset import GPTDatasetConfig
 
         return GPTDatasetConfig(
-            blend=[[str(self.path)], [1.0]],
+            blend=[[str(path) for path in self.paths], self.weights],
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/data/pre_training.py:145]: If self.weights is None & path is a List, can't we just set all the weights to be equal?

File: nemo/lightning/pytorch/callbacks/progress.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def on_train_epoch_start(self, trainer, *_):
         # app_state = AppState()
         # app_state.
 
-        num_training_batches = num_training_batches // calculate_data_parallel_groups()
+        # num_training_batches = num_training_batches // calculate_data_parallel_groups()
```
Comment (marcromeyn) [inline nemo/lightning/pytorch/callbacks/progress.py:40]: We can remove this comment

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 3]
```diff
@@ -37,7 +38,13 @@ def __init__(
         index_mapping_dir: Optional[str] = None,
     ) -> None:
         super().__init__()
-        self.path = path
+        if not isinstance(paths, List):
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/data/pre_training.py:41]:
```suggestion
        if not isinstance(paths,(list, tuple)):
```

File: nemo/collections/llm/gpt/model/base.py [hunk 4]
```diff
@@ -157,7 +157,9 @@ def forward_step(self, batch) -> torch.Tensor:
     def training_step(self, batch, batch_idx=None) -> torch.Tensor:
         # In mcore the loss-function is part of the forward-pass (when labels are provided)
 
-        return self.forward_step(batch)
+        result = self.forward_step(batch)
+        self.log('step', torch.tensor(self.trainer.global_step, dtype=torch.float32))
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/model/base.py:161]: Can't we put this inside the `training_step` of the `MegatronStrategy`?

Comment (marcromeyn) [reply nemo/collections/llm/gpt/model/base.py:161]: That way we can do it once as opposed to each individual model having to do this. The `MegatronStrategy` will wrap the `training_step` here

File: nemo/collections/llm/gpt/data/pre_training.py (hunk 1 above)
Comment (ashors1) [reply nemo/collections/llm/gpt/data/pre_training.py:145]: I tried that initially, but I ran into issues with non-None weights + [num_val_samples=None](https://github.com/NVIDIA/NeMo/blob/d2a86cea28c30678c1c4b1e57beb153ba88d9863/nemo/collections/llm/gpt/data/pre_training.py#L93)

File: nemo/collections/llm/gpt/data/pre_training.py [hunk 5]
```diff
@@ -17,7 +17,8 @@
 class PreTrainingDataModule(pl.LightningDataModule):
```
Comment (jstjohn) [inline nemo/collections/llm/gpt/data/pre_training.py:17]: Could you split out an abstract pl.LightningDataModule that requires `self.data_sampler` to be set, and move the `state_dict` stuff into that? I'm guessing everyone who uses a `data_sampler` and wants to restore the step they were on will need this, right? I don't see anything specific to `weights` etc in those functions and I would like to inherit from a thing that's documented and not have to think about this. What do you think @marcromeyn ?

Comment (jstjohn) [reply nemo/collections/llm/gpt/data/pre_training.py:17]: Maybe it could live in nemo/lightning/data.py and then you could make `PreTrainingDataModule` inherit from that?

Comment (jstjohn) [reply nemo/collections/llm/gpt/data/pre_training.py:17]: Maybe something like this:
```
import pytorch_lightning as pl
from abc import ABC, abstractmethod

class NeMoDataModule(pl.DataModule, ABC):
    def __init__(self):
        super().__init__()
        self._data_sampler = None
        self.initialize_data_sampler()
        if self._data_sampler is None:
            raise ValueError("Child class must set self.data_sampler")

    @abstractmethod
    def initialize_data_sampler(self):
        """Method that must be implemented by the child class to set self.data_sampler."""
        pass

    @property
    def data_sampler(self):
        if self._data_sampler is None:
            raise ValueError("data_sampler must be set")
        return self._data_sampler

    @data_sampler.setter
    def data_sampler(self, value):
        self._data_sampler = value

class PreTrainingDataModule(NeMoDataModule):
    def __init__(self):
        super().__init__()
        # rest of setup, set state needed for later data_sampler initialization
    def initialize_data_sampler(self):
        # NOTE: this is called in the parent class during `super().__init__()`
        self.data_sampler = ... # initialize data sampler here 
```

Comment (marcromeyn) [reply nemo/collections/llm/gpt/data/pre_training.py:17]: Makes sense, let me do that in a future PR

**Review by marcromeyn — APPROVED** [review]
